File: Explore/Learn/Introduction to Data Structure/Binary Tree/Serialize and Deserialize Binary Tree/SerializeDeserialize.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return null;
        
        StringBuilder str = new StringBuilder();
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()){
            TreeNode node = queue.remove();
            
            if (node != null) {
                queue.add(node.left);
                queue.add(node.right);
                str.append(String.valueOf(node.val));
            } else {
                str.append("null");
            }
            
            str.append(",");
        }
        
        str.deleteCharAt(str.length()-1);
        return str.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data == null) return null;
        
        String[] strArr = data.split(",");
        if (strArr[0].equals("null")) return null;
        
        int index = 0;
        TreeNode root = new TreeNode(Integer.valueOf(strArr[0]));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.remove();
            
            int leftIndex = index * 2 + 1;
            int rightIndex = index * 2 + 2;
            
            if (leftIndex < strArr.length) {
                String curr = strArr[leftIndex];
                if (!curr.equals("null")) {
                    TreeNode left = new TreeNode(Integer.parseInt(curr));
                    node.left = left;
                    queue.add(left);
                }
            }
            
            if (rightIndex < strArr.length) {
                String curr = strArr[rightIndex];
                if (!curr.equals("null")) {
                    TreeNode right = new TreeNode(Integer.parseInt(curr));
                    node.right = right;
                    queue.add(right);
                }
            }
            
            index++;
        }
        
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: Explore/Learn/Introduction to Data Structure/Binary Tree/Serialize and Deserialize Binary Tree/SerializeDeserialize.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
	// Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        if (root == null) return ""
        
        var str = StringBuilder()
        val queue = mutableListOf<TreeNode?>()
        queue.add(root)
        
        while (queue.size > 0) {
            val node = queue.removeAt(0)
            
            if (node != null) {
                queue.add(node!!.left)
                queue.add(node!!.right)
            }
            str.append(node?.`val`)
            str.append(",")
        }
        
        str.deleteCharAt(str.length-1)
        return str.toString()
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        if (data == null) return null
        
        var strArr = data.split(",")
        if (strArr[0].equals("null") || strArr[0].equals("")) return null
        
        var index = 0
        var root: TreeNode? = TreeNode(strArr[0].toInt())
        val queue = mutableListOf<TreeNode?>()
        queue.add(root)
        while (queue.size > 0) {
            var node: TreeNode? = queue.removeAt(0)
            
            val leftIndex = index * 2 + 1
            val rightIndex = index * 2 + 2
            
            if (leftIndex < strArr.size) {
                val curr = strArr[leftIndex]
                if (!curr.equals("null")) {
                    var left: TreeNode? = TreeNode(curr.toInt())
                    node?.left = left
                    queue.add(left)
                }
            }
            
            if (rightIndex < strArr.size) {
                val curr = strArr[rightIndex]
                if (!curr.equals("null")) {
                    var right: TreeNode? = TreeNode(curr.toInt())
                    node?.right = right
                    queue.add(right)
                }
            }
            
            index++
        }
        
        return root
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * var ser = Codec()
 * var deser = Codec()
 * var data = ser.serialize(longUrl)
 * var ans = deser.deserialize(data)
 */
